src/hevm/src/EVM/Types.hs

```diff
@@ -123,13 +123,18 @@ data EType
   | End
   deriving (Typeable)
 
+-- Variables refering to a global environment
+newtype GVar (a :: EType) = Id Int
+  deriving (Show, Eq, Ord)
+
 -- add type level list of constraints
 data Expr (a :: EType) where
 
   -- identifiers
 
   Lit            :: W256 -> Expr EWord
   Var            :: Text -> Expr EWord
+  GVar           :: GVar a -> Text -> Expr a
```

Comment:
curious why we need an id and a stringified name here?

Reply:
Yeah I think I would prefer either of the above two. I think both are probably fine. I think you  should be able to use [`eqT`](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Typeable.html#v:eqT)  for type level reflection. 

There is an example in act [here](https://hackage.haskell.org/package/base-4.17.0.0/docs/Data-Typeable.html#v:eqT)

Reply:
I ended up implementing the second bullet